packages/core/src/new-api/internal/execution/execution-engine.ts

```diff
@@ -0,0 +1,219 @@
+  ExecutionStatus,
+} from "../../types/execution-state";
+import {
+  ExecutionResult,
+  FutureRestart,
+  FutureStart,
+  JournalableMessage,
+} from "../../types/journal";
+import {
+  AccountRuntimeValue,
+  Future,
+  FutureType,
+  RuntimeValueType,
+} from "../../types/module";
+import { safeToString } from "../reconciliation/utils";
+
+import { executionStateReducer } from "./executionStateReducer";
+import { isExecutionResult, isOnChainAction, isOnchainResult } from "./guards";
+
+type ExecutionBatch = Future[];
+
+export class ExecutionEngine {
+  public async execute(state: ExecutionEngineState): Promise<DeploymentResult> {
+    const { batches, module } = state;
+
+    const futures = module.getFutures();
+
+    for (const batch of batches) {
+      // TODO: consider changing batcher to return futures rather than ids
+      const executionBatch = batch.map((futureId) =>
+        this._lookupFuture(futures, futureId)
+      );
+
+      const batchResult = await this._executeBatch(executionBatch, state);
+
+      if (batchResult.some((b) => b.type === "execution-failure")) {
+        return { status: "failed" };
+      }
+
+      if (batchResult.some((b) => b.type === "execution-hold")) {
+        return { status: "hold" };
+      }
+
+      if (batchResult.every((b) => b.type !== "execution-success")) {
+        throw new IgnitionError("Unexpected state");
+      }
+    }
+
+    return {
+      status: "success",
+      contracts: this._resolveDeployedContractsFrom(state),
+    };
+  }
+
+  private async _executeBatch(
+    batch: ExecutionBatch,
+    state: ExecutionEngineState
+  ) {
+    return Promise.all(
+      batch.map((entry) => this._executeBatchEntry(entry, state))
+    );
+  }
+
+  private async _executeBatchEntry(
+    future: Future,
+    state: ExecutionEngineState
+  ): Promise<ExecutionResult> {
+    let current: JournalableMessage = this._startOrRestartFor(future, state);
+    await this._apply(state, current);
+
+    const context = {
+      executionState: state.executionStateMap[future.id],
+    };
+
+    const exectionStrategy = state.strategy.executeStrategy(context);
+
+    while (!isExecutionResult(current)) {
+      context.executionState = state.executionStateMap[future.id];
+
+      if (isOnChainAction(current)) {
+        current = await state.transactionService.onchain(current);
+      } else if (isOnchainResult(current)) {
+        current = (await exectionStrategy.next(current)).value;
+      } else {
+        current = (await exectionStrategy.next(null)).value;
+      }
+
+      await this._apply(state, current);
+    }
+
+    return current;
+  }
+
+  private _startOrRestartFor(
+    future: Future,
+    state: ExecutionEngineState
+  ): FutureStart | FutureRestart {
+    const executionState = state.executionStateMap[future.id];
+
+    if (executionState === undefined) {
+      return this._initCommandFor(future, state);
+    }
+
+    return { type: "execution-restart", futureId: future.id };
```

Comment:
What does restart mean here?

Reply:
This might not make sense.

I want us to get to restarting a future in the transaction service. I am not sure as whether the transaction service detects the restart or we do it based on the previously recorded execution state.